File: LeetCode/108.Convert_Sorted_Array_to_Binary_Search_Tree/108.Convert_Sorted_Array_to_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        // create tree
        TreeNode root = createTree(nums, 0, nums.length - 1);
        return root;
    }

    private TreeNode createTree(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        int index = (end + start) / 2;
        int num = nums[index];
        TreeNode node = new TreeNode(num);
        node.left = createTree(nums, start, index - 1);
        node.right = createTree(nums, index + 1, end);
        return node;
    }
}

File: LeetCode/108.Convert_Sorted_Array_to_Binary_Search_Tree/108.Convert_Sorted_Array_to_Binary_Search_Tree.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:

        def numsToBST(l, r):
            if l > r:
                return None
            if l == r:
                return TreeNode(nums[l])

            mid = (l + r) // 2
            left = numsToBST(l, mid - 1)
            right = numsToBST(mid + 1, r)

            node = TreeNode(nums[mid])
            node.left, node.right = left, right
            return node

        return numsToBST(0, len(nums) - 1)
